File: apps/worker/src/checkers/ping.ts
import type { Monitor } from "@unstatus/db";
import { connect, type Socket } from "net";
import { assertPublicHostTarget } from "./egress.js";

const RETRY_DELAY_MS = 1500;

type PingResult = {
  status: string;
  latency: number;
  statusCode: null;
  message: string | null;
  responseHeaders: null;
  responseBody: null;
};

/**
 * Ping check using TCP connect (SYN) to common web ports.
 * We probe 443 first, then 80 if the host explicitly refuses the TLS port.
 */
async function probePort(host: string, port: number, timeoutMs: number) {
  const startedAt = performance.now();

  return new Promise<{ reachable: boolean; refused: boolean; message: string | null; latency: number }>((resolve) => {
    let socket: Socket | null = null;
    let settled = false;

    const finish = (result: { reachable: boolean; refused: boolean; message: string | null }) => {
      if (settled) return;
      settled = true;
      socket?.destroy();
      resolve({
        ...result,
        latency: Math.round(performance.now() - startedAt),
      });
    };

    socket = connect({ host, port });
    socket.setTimeout(timeoutMs);

    socket.once("connect", () => {
      finish({ reachable: true, refused: false, message: null });
    });

    socket.once("timeout", () => {
      finish({ reachable: false, refused: false, message: "Connection timed out" });
    });

    socket.once("error", (err: Error) => {
      const msg = err.message;

      if (msg.includes("ECONNRESET")) {
        finish({ reachable: true, refused: false, message: null });
        return;
      }

      finish({
        reachable: false,
        refused: msg.includes("ECONNREFUSED"),
        message: formatPingError(err),
      });
    });
  });
}

async function performPing(monitor: Monitor): Promise<PingResult> {
  const startedAt = performance.now();
  const rawHost = monitor.host ?? monitor.url ?? "";
  const timeoutMs = monitor.timeout * 1000;

  if (!rawHost) {
    return {
      status: "down",
      latency: 0,
      statusCode: null,
      message: "No host configured",
      responseHeaders: null,
      responseBody: null,
    };
  }

  let host: string;
  try {
    host = await assertPublicHostTarget(rawHost);
  } catch (e) {
    return {
      status: "down",
      latency: Math.round(performance.now() - startedAt),
      statusCode: null,
      message: e instanceof Error ? e.message : "Target host is not allowed",
      responseHeaders: null,
      responseBody: null,
    };
  }

  const primary = await probePort(host, 443, timeoutMs);
  if (primary.reachable) {
    return {
      status: "up",
      latency: primary.latency,
      statusCode: null,
      message: null,
      responseHeaders: null,
      responseBody: null,
    };
  }

  if (primary.refused) {
    const fallback = await probePort(host, 80, timeoutMs);
    return {
      status: fallback.reachable ? "up" : "down",
      latency: fallback.latency,
      statusCode: null,
      message: fallback.reachable ? null : fallback.message,
      responseHeaders: null,
      responseBody: null,
    };
  }

  return {
    status: "down",
    latency: primary.latency,
    statusCode: null,
    message: primary.message,
    responseHeaders: null,
    responseBody: null,
  };
}

export async function checkPing(monitor: Monitor) {
  const first = await performPing(monitor);
  if (first.status !== "down") return first;

  await new Promise((r) => setTimeout(r, RETRY_DELAY_MS));
  return performPing(monitor);
}

function formatPingError(err: Error): string {
  const msg = err.message;
  if (msg.includes("ENOTFOUND") || msg.includes("getaddrinfo")) return "DNS resolution failed";
  if (msg.includes("ECONNREFUSED")) return "Host refused connection";
  if (msg.includes("ETIMEDOUT")) return "Connection timed out";
  if (msg.includes("EHOSTUNREACH")) return "Host unreachable";
  if (msg.includes("ENETUNREACH")) return "Network unreachable";
  return msg;
}

File: apps/worker-rs/src/checks/ping.rs
use std::io::ErrorKind;
use std::time::Duration;

use tokio::net::TcpStream;
use tokio::time::{sleep, timeout};

use crate::models::{CheckResult, WorkerMonitorRow};

const RETRY_DELAY_MS: u64 = 1_500;

pub async fn check_ping(monitor: &WorkerMonitorRow) -> anyhow::Result<CheckResult> {
    let first = perform_ping(monitor).await;
    if first.status != "down" {
        return Ok(first);
    }

    sleep(Duration::from_millis(RETRY_DELAY_MS)).await;
    Ok(perform_ping(monitor).await)
}

async fn perform_ping(monitor: &WorkerMonitorRow) -> CheckResult {
    let Some(host) = normalize_host(monitor) else {
        return CheckResult {
            status: "down".to_string(),
            latency: 0,
            status_code: None,
            message: Some("No host configured".to_string()),
            response_headers: None,
            response_body: None,
        };
    };

    let timeout_duration = Duration::from_secs(monitor.timeout.max(1) as u64);
    let first_attempt = connect_host(&host, 443, timeout_duration).await;
    if first_attempt.status == "up" {
        return first_attempt;
    }

    if matches!(
        first_attempt.message.as_deref(),
        Some("Host refused connection")
    ) {
        let second_attempt = connect_host(&host, 80, timeout_duration).await;
        if second_attempt.status == "up" {
            return second_attempt;
        }
    }

    first_attempt
}

async fn connect_host(host: &str, port: u16, timeout_duration: Duration) -> CheckResult {
    let start = std::time::Instant::now();
    let result = timeout(timeout_duration, TcpStream::connect((host, port))).await;
    let latency = start.elapsed().as_millis() as i32;

    match result {
        Ok(Ok(_)) => CheckResult {
            status: "up".to_string(),
            latency,
            status_code: None,
            message: None,
            response_headers: None,
            response_body: None,
        },
        Ok(Err(error)) => CheckResult {
            status: "down".to_string(),
            latency,
            status_code: None,
            message: Some(format_ping_error(&error)),
            response_headers: None,
            response_body: None,
        },
        Err(_) => CheckResult {
            status: "down".to_string(),
            latency,
            status_code: None,
            message: Some("Connection timed out".to_string()),
            response_headers: None,
            response_body: None,
        },
    }
}

fn normalize_host(monitor: &WorkerMonitorRow) -> Option<String> {
    let raw = monitor
        .host
        .clone()
        .or_else(|| monitor.url.clone())
        .unwrap_or_default();

    let stripped = raw
        .trim()
        .trim_start_matches("http://")
        .trim_start_matches("https://");
    let host = stripped
        .split('/')
        .next()
        .unwrap_or("")
        .split(':')
        .next()
        .unwrap_or("");
    (!host.is_empty()).then(|| host.to_string())
}

fn format_ping_error(error: &std::io::Error) -> String {
    match error.kind() {
        ErrorKind::NotFound => "DNS resolution failed".to_string(),
        ErrorKind::ConnectionRefused => "Host refused connection".to_string(),
        ErrorKind::TimedOut => "Connection timed out".to_string(),
        ErrorKind::AddrNotAvailable => "Host unreachable".to_string(),
        ErrorKind::NetworkUnreachable => "Network unreachable".to_string(),
        _ => error.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use crate::models::WorkerMonitorRow;

    use super::normalize_host;

    #[test]
    fn normalize_host_strips_scheme_and_path() {
        let monitor = WorkerMonitorRow {
            id: "monitor_1".to_string(),
            organization_id: "org_1".to_string(),
            name: "API".to_string(),
            monitor_type: "ping".to_string(),
            interval: 60,
            timeout: 10,
            url: Some("https://example.com/status?x=1".to_string()),
            method: None,
            headers: None,
            body: None,
            host: None,
            port: None,
            rules: None,
            regions: serde_json::json!(["eu"]),
            auto_incidents: false,
            last_checked_at: None,
            next_check_at: None,
        };

        assert_eq!(normalize_host(&monitor).as_deref(), Some("example.com"));
    }
}
